src/python/pants/build_graph/target.py

```diff
@@ -32,6 +33,47 @@
 logger = logging.getLogger(__name__)
 
 
+class SyntheticTargetNotFound(Exception):
+  pass
+
+
+def _get_synthetic_target(target, thrift_dep):
+  """Find a thrift_dep's corresponding synthetic target in the dependencies of the given target."""
+  for dep in target.dependencies:
+    if dep != thrift_dep and dep.is_synthetic and dep.derived_from == thrift_dep:
+      return dep
+  raise SyntheticTargetNotFound('No synthetic target is found for thrift target: {}'.format(thrift_dep))
```

Comment:
As noted above, this is a bit broken. For one `s/thrift_dep/codegen_dep/`, but can this all be solved by returning `thrift_dep` here? IE: change this function to return a derived dependency if one exists; otherwise, return the raw dependency.

Reply:
Yes: good point. But this whole thing is a hack, and I think there is a cleaner way to solve it. Will comment and punt for #4936, which I think we can fix during #4677.